File: apoyo.cpp
#include <iostream>
#include "apoyo.h"
using namespace std;
//----------------------------
// Definir las direcciones: {izquierda, derecha, arriba, abajo}
//------------------------------
objeto::objeto() : tipo(VACIO), x(0), y(0) {}  // Inicializamos como VACIO por defecto
objeto::objeto(int posx, int posy, int dirx, int diry, tipoObjeto _tipo) : x(posx), y(posy),direccion_x(dirx),direccion_y(diry),tipo(_tipo) {}
tablero::tablero(int f, int c) {
    filas = f;
    columnas = c;
    tamano = f * c;

    // Asignación de memoria para la matriz de punteros a objetos
    matriz = new objeto*[filas];
    for (int i = 0; i < filas; ++i) {
        matriz[i] = new objeto[columnas];  // Cada fila es un array de objetos
    }

    // Inicialización de cada elemento de la matriz como un objeto VACIO
    for (int i = 0; i < filas; i++) {
        for (int j = 0; j < columnas; j++) {
            matriz[i][j] = objeto();  // Inicialización por defecto (VACIO)
        }
    }
}

tablero ::~tablero() {
            for (int i = 0; i < filas; ++i) {
                delete[] matriz[i];
            }
            delete[] matriz;
        }
bool tablero::detectar_choque(int x, int y, int &dx, int &dy) {
        if (matriz[x][y].tipo == ESPEJO) {
            cout << "La luz ha chocado con un espejo en (" << x << ", " << y << ")" << endl; //ayuda para indicar si ha chocado o no
            dx = matriz[x][y].direccion_x;
            dy= matriz[x][y].direccion_y;
            return false;
        } else if (matriz[x][y].tipo == OBSTACULO) {
            cout << "La luz ha chocado con un obstáculo en (" << x << ", " << y << ")" << endl; //ayuda para indicar si ha chocado o no
            dx = matriz[x][y].direccion_x;
            dy= matriz[x][y].direccion_y;
            return true;
        }
    }
void tablero::colocar_objeto(int posx, int posy, int dirx, int diry, tipoObjeto _tipo){
        if(_tipo == OBSTACULO){
            objeto OBJETO(posx,posy,0,0,_tipo);
            matriz[OBJETO.x][OBJETO.y]= OBJETO;
        }
        else{
            objeto OBJETO(posx,posy,dirx,diry,_tipo);
            matriz[OBJETO.x][OBJETO.y]= OBJETO;
        }
}
void tablero::mostrar_tablero(){
    for (int i = 0; i < filas; i++) {
        for (int j = 0; j < columnas; j++) {
            cout << "|";
            cout.flush();
            if (matriz[i][j].tipo == ESPEJO) cout << "E "; // espejo
            else if (matriz[i][j].tipo == OBSTACULO) cout << "OB"; // obstáculo
            else if (matriz[i][j].tipo == VACIO) cout << "  "; // espacio vacío
            cout << "|";
        }
        cout << endl;
    }

}

//-------------------------------------------------------------------------------------------------------------
    //Creacion de objeto "luz"
luz::luz(int posx, int posy,int dirx,int diry,tipoObjeto _tipo) : objeto(posx, posy, dirx, diry, LUZ) {}
void luz::viajar_luz(tablero mapa){
        if (mapa.detectar_choque(x, y,direccion_x,direccion_y)==true) {//choca con obstaculo
            return ;// Detiene el viaje si hay choque
        }
        // Continuar con las reglas de movimiento
        if (y+1 < mapa.filas and direccion_y == 1 and direccion_x == 0) {
            y = y + 1;
            viajar_luz(mapa);
        }
        else if (y - 1 >= 0 and direccion_y == -1 and direccion_x == 0) {
            y = y - 1;
            viajar_luz(mapa);
        }
        else if (x + 1 < mapa.columnas and direccion_y == 0 and direccion_x == 1) {
            x = x + 1;
            viajar_luz(mapa);
        }
        else if (x - 1 >= 0 and direccion_y == 0 and direccion_x == -1) {
            x = x - 1;
            viajar_luz(mapa);
        }
    }
//---------------------------------------------------------------------------------------
    //Creacion de objeto "espejo"
espejo::espejo(int posx, int posy,int dirx,int diry,tipoObjeto _tipo) : objeto(posx, posy, dirx, diry, ESPEJO) {}
//--------------------------------------------------------------------------------------
    //Creacion de objeto "obstaculo"
obstaculo::obstaculo(int posx, int posy,int dirx,int diry,tipoObjeto _tipo) : objeto(posx, posy, 0, 0, OBSTACULO) {}
//-----------------------------------------------------------------------------------------

File: apoyo.py
import time
import os

# Nota: x = columnas , y  = filas
"""Class Tipos de movimiento: Ilustrativo
    (0,0) ->quieto
    (0,1) ->derecha
    (0,-1) ->izquierda
    (1,0) ->abajo
    (-1,0) ->arriba
    (1,1) ->diagonal derecha hacia abajo
    (1,-1) ->diagonal izquierda hacia abajo
    (-1,1) ->diagonal derecha hacia arriba
    (-1,-1) ->diagonal izquierda hacia arriba
##"""

class TipoObjeto:
    VACIO = 0
    OBSTACULO = 1
    ESPEJO = 2
    LUZ = 3

class TipoEspejo:
    DIAGONAL1 = 1  # Representa `/`
    DIAGONAL2 = 2  # Representa `\`
    LATERAL1 = 3   # Representaa '__'
    LATERAL2 = 4   # Representaa '|'

class Objeto:
    def __init__(self, x, y, tipo):
        self.x = x
        self.y = y
        self.tipo = tipo

class Tablero:
    def __init__(self, filas, columnas):
        self.filas = filas
        self.columnas = columnas
        self.matriz = [[Objeto(x, y, TipoObjeto.VACIO) for y in range(columnas)] for x in range(filas)]

    def colocar_objeto(self, x, y, objeto):
        self.matriz[x][y] = objeto

    def obtener_objeto(self, x, y):
        return self.matriz[x][y]
    
    def mostrar_tablero(self):
        # Limpia la consola antes de mostrar el tablero
        os.system('cls' if os.name == 'nt' else 'clear')
        
        for i in range(self.filas):
            for j in range(self.columnas):
                obj = self.matriz[i][j]
                if obj.tipo == TipoObjeto.ESPEJO:
                    if obj.tipo_espejo == TipoEspejo.DIAGONAL1: 
                        print("|", "/", end="|")
                    elif obj.tipo_espejo == TipoEspejo.DIAGONAL2: 
                        print("|", "\\", end="|")
                    elif obj.tipo_espejo == TipoEspejo.LATERAL1: 
                        print("|", "_", end="|")
                    elif obj.tipo_espejo == TipoEspejo.LATERAL2: 
                        print("|", "|", end="|")
                elif obj.tipo == TipoObjeto.OBSTACULO:
                    print("|", "OB", end="|")
                elif obj.tipo == TipoObjeto.LUZ:
                    print("|", "L", end="|")
                else:
                    print("|", " ", end="|")
            print()
        time.sleep(0.5)  # Pausa de 0.5 segundos para simular movimiento en tiempo real

class Espejo(Objeto):
    def __init__(self, x, y, tipo_espejo):
        super().__init__(x, y, TipoObjeto.ESPEJO)
        self.tipo_espejo = tipo_espejo

class Obstaculo(Objeto):
    def __init__(self, x, y):
        super().__init__(x, y, TipoObjeto.OBSTACULO)

class Luz(Objeto):
                #(px,py, arriba_abajo,izquierda o derecha)
                #direccion = 0, significa que no se mueve
    def __init__(self, x, y, direccion_x, direccion_y):
        super().__init__(x, y, TipoObjeto.LUZ)
        self.direccion_x = direccion_x
        self.direccion_y = direccion_y

    def detectar_choque(self, tablero):
        # Calcular la posición hacia la cual se moverá la luz
        proxima_x = self.x + self.direccion_x
        proxima_y = self.y + self.direccion_y

        # Comprobar si está dentro de los límites del tablero
        if not (0 <= proxima_x < tablero.filas and 0 <= proxima_y < tablero.columnas):
            print(f"La luz ha salido del tablero en la posición ({self.x}, {self.y})")
            return True

        objeto = tablero.obtener_objeto(proxima_x, proxima_y)
        if objeto.tipo == TipoObjeto.ESPEJO:
            print(f"La luz ha chocado con un espejo en ({proxima_y}, {proxima_x})")
            # Reflexión dependiendo del tipo de espejo
            if objeto.tipo_espejo == TipoEspejo.DIAGONAL1:  # `/`
                self.direccion_x, self.direccion_y = -self.direccion_y, -self.direccion_x
            elif objeto.tipo_espejo == TipoEspejo.DIAGONAL2:  # `\`
                self.direccion_x, self.direccion_y = self.direccion_y, self.direccion_x
            elif objeto.tipo_espejo == TipoEspejo.LATERAL1:  # `__`
                self.direccion_x, self.direccion_y = -self.direccion_x, self.direccion_y
            elif objeto.tipo_espejo == TipoEspejo.LATERAL2:  # `|`
                self.direccion_x, self.direccion_y = self.direccion_x, -self.direccion_y
            return False
        elif objeto.tipo == TipoObjeto.OBSTACULO:
            print(f"La luz ha chocado con un obstáculo en ({proxima_y}, {proxima_x})")
            return True
        return False
    
    def viajar_luz(self, tablero):
        while True:
            if self.detectar_choque(tablero):
                break

            # Actualizar la posición de la luz en el tablero
            tablero.colocar_objeto(self.x, self.y, Objeto(self.x, self.y, TipoObjeto.VACIO))  # Limpia la posición actual

            # Mueve la luz en la dirección asignada
            self.x += self.direccion_x
            self.y += self.direccion_y

            # Coloca la luz en la nueva posición y muestra el tablero
            tablero.colocar_objeto(self.x, self.y, self)
            tablero.mostrar_tablero()

            # Pausa para visualización
            time.sleep(0.5)
